README.md
---
page_type: sample
languages:
  - python
products:
  - Azure Video Indexer
  - Azure Search
---
[![CI](https://github.com/Azure-Samples/azureVISearchableContent/actions/workflows/ci.yml/badge.svg?branch=main)](https://github.com/Azure-Samples/azureVISearchableContent/actions/workflows/ci.yml)

# Azure Video Analyzer Searchable Moments

This project reads [Azure Video Indexer](https://azure.microsoft.com/en-in/services/media-services/video-indexer/) output file (JSON format) into small  chunks of searchable content.

So you can make the following queries to the azure search index to find the right moments of a video:

- _"Find the moments where '**Jennifer Aniston**' in Friends TV show is happy"_

```json
{
  "search": "Jennifer Aniston",
  "searchFields": "faces/face, name",
  "filter": "search.ismatchscoring('friends')and (emotions/any(em: em/emotion eq 'Joy')) "
}
```

- _"Find the moments in any show where the transcript is '**getting old**'"_
```json
{
  "search": "getting old",
  "searchFields": "transcripts/transcript"
}
```
- _"Find the moments in Friends show where '**Jean Claude**' Appears_"
```json
{  
  "search": "Jean Claude",
  "searchFields": "faces/face, namedPeople/namedPerson",
  "filter": "search.ismatchscoring('friends') ",
}
```

## Problem Summary

With the vast growth and progress of multimedia the Internet, particularly videos, text-based retrieval search engines plays an important role on searchability of videos.

With Azure Video Analyzer, developers can quickly build an AI-powered video analytics solution to extract actionable insights from videos, whether stored or streaming. Organizations can enhance workplace safety, in-store experiences, digital asset management, media archives, and more by understanding audio and visual video content without video nor ML expertise.

Azure Video Analyzer produces one flat json file of a provided video that covers a full video duration. How we can come up with a search index schema which is capable of retrieving data that is most relevant to a moment of a video?


## Solution Summary

Azure In this repository we are processing the output of the Video Analyzer into moments (n seconds) where each moment covers n seconds of a video. We create a Search index having all the features extracted from the Video Analyzer and we upload the extracted moments into the search index. Finally, providing a search query, we can retrieve the related moments of a video from the Azure Search Index.
The Search index has flexibility of having more fields to accommodate custom features (e.g., features not existing in Azure Video Analyzer and created for that custom usage).


## Products/Technologies/Languages Used

- Products & Technologies:
    - Azure Video analyzer
    - Azure Search
    - Azure Storage Account (optional)
    - Docker (Optional)
- Languages:
    - Python

## Architecture (Optional)

- In [this documentation page](Architecture.md) you may find the architecture of the Azure Video Indexer Searchable moments.

## Getting Started

The project can be run like any other Python code, or packaged into a Docker image and run
accordingly.

Below we detail both approaches.

### General set up

For both local development and running the code without Docker, you'll need to do the following.

**Create a virtual environment**

On Mac and linux:
`python3 -m venv env`

On Windows:
`py -m venv env`

**Activate your virtual environment**

On macOS and Linux:
`source env/bin/activate`

On Windows:

`.\env\Scripts\activate`

**Add required libraries**

`pip install -r src/requirements.txt`

### Run the project

#### Standard - From YAML config

Copy the contents of [`config-dev.yml`](src/config/config-dev.yml) to a new gitignored file
`config.yml`, and add the required details. NOTE: These include secrets and so should never be
pushed to version control.

```yaml
# this is used when you want to parse files from storage account
storage:
  connection-string: "YOUR STORAGE ACCOUNT CONNECTION STRING"
  # a container which you store JSON files processed by Azure Video Indexer
  container: "YOUR CONTAINER NAME"

search:
  service-name: "SEARCH SERVICE NAME"
  api-version: "SEARCH SERVICE API VERSION"
  api-key: "SEARCH SERVICE API KEY"
  index-name:  "NAME OF THE SEARCH INDEX YOU WANT TO CREATE"
  index-schema-path: "PATH TO SEARCH INDEX SCHEMA e.g. client/index-schema.json"

files:
  vi-output-directory: "PATH TO VI JSON FILES e.g. client/files/vi-files"
  processed-directory: "PATH TO PROCESSED FILE DIRECTORY e.g. client/files/processed"
  ingested-file: "NAME OF THE LOGS FOR PROCESSED ASSETS e.g. ingested.txt"
  failed-to-ingest-file: "NAME OF THE LOGS FOR FAIL PROCESSED ASSETS e.g. failed-to-ingest.txt"

# interval durations you wish to create in milliseconds
parser:
  milliseconds-interval: 10000
```

Navigate into the `src` directory with `cd src`.

Now using command line you can run the scripts in two ways:

1. Process insights files from your specified storage account and container: `python main.py`
2. Process insights files from the local filesystem: `python main.py local`


#### Alternative - Using Docker

When using Docker, we build and run a Python image, passing in the config details as environment
variables, and have the option of mounting our index schema and any other files needed from the
host (local machine).

First [ensure you have Docker installed and working](https://docs.docker.com/get-docker/).

As your `config.yml` will not be configurable once an image is built, remove it if created, or just
leave the repo in its original state. With only `config-dev.yml`, the code will rely upon relevant
environment variables passed in at runtime to determine how it should run.

From the root of the repository, build the docker image using:

```bash
docker -t azure-vi-searchable-content .
```

See the [`Dockerfile`](Dockerfile) for details on how the image is constructed.

To run the image successfully, we need to pass in environment variables as follows:

```bash
docker run \
-e INSIGHTS_CONTAINER_NAME="YOUR INSIGHTS CONTAINER NAME" \                                                                                                                                                                                                   ✹
-e SEARCH_SERVICE_NAME="YOUR SEARCH SERVICE NAME" \
-e SEARCH_API_VERSION="2020-06-30" \
-e SEARCH_API_KEY="YOUR SEARCH API KEY" \
-e SEARCH_INDEX_NAME="YOUR DESIRED SEARCH INDEX NAME" \
-e INDEX_SCHEMA_PATH="client/index-schema.json" \
-e VI_OUTPUT_DIRECTORY="client/files/vi-files" \
-e FILE_PROCESSING_LOGS_DIR="client/files/processed" \
-e INGEST_LOG_FILENAME="ingested.txt" \
-e INGEST_FAILURE_LOG_FILENAME="failed-to-ingest.txt" \
-e MILLISECONDS_INTERVAL=10000 \
-e STORAGE_CONNECTION_STRING="YOUR CONNECTION STRING" \
azure-vi-searchable-content:latest
```

Note that this will still retrieve your search index schema and write log files to inside the
container. This can be undesirable, so we can make all these configurable on the host machine by
volume mounting the `mount-files` directory, and managing the files there.

```bash
docker run \
-e INSIGHTS_CONTAINER_NAME="YOUR INSIGHTS CONTAINER NAME" \                                                                                                                                                                                                   ✹
-e SEARCH_SERVICE_NAME="YOUR SEARCH SERVICE NAME" \
-e SEARCH_API_VERSION="2020-06-30" \
-e SEARCH_API_KEY="YOUR SEARCH API KEY" \
-e SEARCH_INDEX_NAME="YOUR DESIRED SEARCH INDEX NAME" \
-e INDEX_SCHEMA_PATH="mount-files/index-schema.json" \
-e VI_OUTPUT_DIRECTORY="mount-files/logs" \
-e FILE_PROCESSING_LOGS_DIR="mount-files/logs" \
-e INGEST_LOG_FILENAME="ingested.txt" \
-e INGEST_FAILURE_LOG_FILENAME="failed-to-ingest.txt" \
-e MILLISECONDS_INTERVAL=10000 \
-e STORAGE_CONNECTION_STRING="YOUR CONNECTION STRING" \
-v "$(pwd)"/mount-files:/src/mount-files \
azure-vi-searchable-content:latest
```

When using the command above, changes to your index schema will be used in a newly running
container, without need to rebuild the image.

This can be simplified further using a gitignored .env file ([`.env-dev`](./.env-dev) is given as
an example), running the image as follows:

```bash
docker run \
--env-file "./.env" \
-v "$(pwd)"/mount-files:/src/mount-files \
azure-vi-searchable-content:latest
```

#### How it works - client and parser usage

**1. Read data from the blob:**

To read  the JSON files from the blob:

- Make sure the [config.yml](src/config/config-dev.yml) has required values to connect to the storage account.
- Run the [storage-client.py](src/client/storageClient.py) script

This will read the data from a container into your local machine


**2. Create/Upload data into Azure Search:**
Now that you have the data in your local machine, you can create a search index and after parsing those data, upload them into the search index

Run [search-client.py](src/client/searchClient.py) which performs the following:

- Creates a search index using the file `index-schema.json` ([here](src/client/index-schema.json)
  as standard and [here](mount-files/index-schema.json) for Docker volume mounting)
- Reads JSON files from `files/blob-files`
- Using Parser class parses each json file and creates list of created intervals
- Upload the data into azure search index

**Note 1:** By default (filename is configurable) every successful upload will be logged to  `<Your-specified-logs-directory>/ingested.txt`

**Note 2:** By default (filename is configurable) every unsuccessful upload will be logged to `<Your-specified-logs-directory>/failed-to-ingest.txt`

## Related resources

- [Azure Video Analyzer](https://azure.microsoft.com/en-us/products/video-analyzer/)
- [Create Azure Search Index](https://docs.microsoft.com/en-us/rest/api/searchservice/create-index)


CHANGELOG.md
## Azure Video Analyzer Searchable Moments Changelog

<a name="x.y.z"></a>
# Initial Integration into Azure Samples  (2022-02-07)

*Features*
* Create a search index having all the features existing in Azure Video Analyzer and having the flexibility to have more fields as Custome feature.

* Parse the Azure Video Analyzer json output file into small intervals (n seconds) and feed into the azure search index 

tests/utils.py
import json
import os


class Utils:
    def read_json_from_resources(self, path, change_key_to_number=False):
        try:
            path = os.path.join("tests/resources", path)
            with open(path) as f:

                json_file = json.load(f)
                if change_key_to_number:
                    json_file = self.convert_json_keys_to_int(json_file)
                return json_file
        except ValueError:
            print("could not read from the json file")

    @staticmethod
    def convert_json_keys_to_int(dict_with_string_keys):
        dict_with_int_keys = dict()
        for k, v in dict_with_string_keys.items():
            dict_with_int_keys[int(k)] = v
        return dict_with_int_keys


tests/testparser.py
import os

from src.parser.parser import Parser
from tests.testbase import TestBase
from tests.utils import Utils


class TestParser(TestBase):
    """
    This class contains unit tests for UsernameHelper class
    """
    os.environ["MILLISECONDS_INTERVAL"] = "10000"
    parser = Parser()
    utils = Utils()

    def test_create_intervals(self):
        """
        test_alphanum_split for this pattern: "alpha_number_alpha".
        """
        # GIVEN
        vi_output = self.utils.read_json_from_resources("vi-output.json")
        video = vi_output["videos"][0]
        # WHEN
        actual = self.parser.create_intervals(video, "FOO")
        # THEN
        self.assert_equals(
            actual, self.utils.read_json_from_resources("intervals.json", True)
        )

    def test_insert_items_to_intervals(self):
        """
        test_alphanum_split for this pattern: "alpha_number_alpha".
        """
        # GIVEN
        expected = {
            0: {"bar_0": {}, "bar_2": [{"FOO": "BAZ"}]},
            10000: {"bar_1": {}, "bar_2": [{"FOO": "BAZ"}]},
            200000: {"bar_2": {}},
            30000: {"bar_3": {}},
            40000: {"bar_4": {}},
        }
        occurred_intervals = [0, 10000]
        item_tuple = ("bar_2", {"FOO": "BAZ"})
        intervals = {
            0: {"bar_0": {}},
            10000: {"bar_1": {}},
            200000: {"bar_2": {}},
            30000: {"bar_3": {}},
            40000: {"bar_4": {}},
        }
        # WHEN
        actual = self.parser.insert_items_to_intervals(
            occurred_intervals, item_tuple, intervals
        )
        # THEN
        self.assert_equals(actual, expected)

    def test_parse_transcript(self):
        # GIVEN
        vi_output = self.utils.read_json_from_resources("vi-output.json")
        transcript = vi_output["videos"][0]["insights"]["transcript"]
        intervals = self.parser.create_intervals(vi_output["videos"][0], "FOO")
        # WHEN
        actual = self.parser.parse_transcript(transcript, intervals)
        # THEN
        self.assert_equals(
            actual,
            self.utils.read_json_from_resources("intervals-with-transcript.json", True),
        )

    def test_parse_ocr(self):
        # GIVEN
        vi_output = self.utils.read_json_from_resources("vi-output.json")
        ocr = vi_output["videos"][0]["insights"]["ocr"]
        intervals = self.parser.create_intervals(vi_output["videos"][0], "FOO")
        # WHEN
        actual = self.parser.parse_ocr(ocr, intervals)
        # THEN
        self.assert_equals(
            actual, self.utils.read_json_from_resources("intervals-with-ocr.json", True)
        )

    def test_parse_keywords(self):
        # GIVEN
        vi_output = self.utils.read_json_from_resources("vi-output.json")
        keyword = vi_output["videos"][0]["insights"]["keywords"]
        intervals = self.parser.create_intervals(vi_output["videos"][0], "FOO")
        # WHEN
        actual = self.parser.parse_keywords(keyword, intervals)
        # THEN
        self.assert_equals(
            actual,
            self.utils.read_json_from_resources("intervals-with-keyword.json", True),
        )

    def test_parse_topics(self):
        # GIVEN
        vi_output = self.utils.read_json_from_resources("vi-output.json")
        topics = vi_output["videos"][0]["insights"]["topics"]
        intervals = self.parser.create_intervals(vi_output["videos"][0], "FOO")
        # WHEN
        actual = self.parser.parse_topics(topics, intervals)
        # THEN
        self.assert_equals(
            actual,
            self.utils.read_json_from_resources("intervals-with-topics.json", True),
        )

    def test_parse_faces(self):
        # GIVEN
        vi_output = self.utils.read_json_from_resources("vi-output.json")
        faces = vi_output["videos"][0]["insights"]["faces"]
        intervals = self.parser.create_intervals(vi_output["videos"][0], "FOO")
        # WHEN
        actual = self.parser.parse_faces(faces, intervals)
        # THEN
        self.assert_equals(
            actual,
            self.utils.read_json_from_resources("intervals-with-faces.json", True),
        )

    def test_parse_labels(self):
        # GIVEN
        vi_output = self.utils.read_json_from_resources("vi-output.json")
        labels = vi_output["videos"][0]["insights"]["labels"]
        intervals = self.parser.create_intervals(vi_output["videos"][0], "FOO")
        # WHEN
        actual = self.parser.parse_labels(labels, intervals)
        # THEN
        self.assert_equals(
            actual,
            self.utils.read_json_from_resources("intervals-with-labels.json", True),
        )

    def test_parse_named_locations(self):
        # GIVEN
        vi_output = self.utils.read_json_from_resources("vi-output.json")
        named_locations = vi_output["videos"][0]["insights"]["namedLocations"]
        intervals = self.parser.create_intervals(vi_output["videos"][0], "FOO")
        # WHEN
        actual = self.parser.parse_named_locations(named_locations, intervals)
        # THEN
        self.assert_equals(
            actual,
            self.utils.read_json_from_resources(
                "intervals-with-named-locations.json", True
            ),
        )

    def test_parse_named_people(self):
        # GIVEN
        vi_output = self.utils.read_json_from_resources("vi-output.json")
        named_people = vi_output["videos"][0]["insights"]["namedPeople"]
        intervals = self.parser.create_intervals(vi_output["videos"][0], "FOO")
        # WHEN
        actual = self.parser.parse_named_people(named_people, intervals)
        # THEN
        self.assert_equals(
            actual,
            self.utils.read_json_from_resources(
                "intervals-with-named-people.json", True
            ),
        )

    def test_parse_audio_effects(self):
        # GIVEN
        vi_output = self.utils.read_json_from_resources("vi-output.json")
        audio_effects = vi_output["videos"][0]["insights"]["audioEffects"]
        intervals = self.parser.create_intervals(vi_output["videos"][0], "FOO")
        # WHEN
        actual = self.parser.parse_audio_effects(audio_effects, intervals)
        # THEN
        self.assert_equals(
            actual,
            self.utils.read_json_from_resources(
                "intervals-with-audio-effects.json", True
            ),
        )

    def test_parse_sentiments(self):
        # GIVEN
        vi_output = self.utils.read_json_from_resources("vi-output.json")
        sentiments = vi_output["videos"][0]["insights"]["sentiments"]
        intervals = self.parser.create_intervals(vi_output["videos"][0], "FOO")
        # WHEN
        actual = self.parser.parse_sentiments(sentiments, intervals)
        # THEN
        self.assert_equals(
            actual,
            self.utils.read_json_from_resources("intervals-with-sentiments.json", True),
        )

    def test_parse_emotions(self):
        # GIVEN
        vi_output = self.utils.read_json_from_resources("vi-output.json")
        emotions = vi_output["videos"][0]["insights"]["emotions"]
        intervals = self.parser.create_intervals(vi_output["videos"][0], "FOO")
        # WHEN
        actual = self.parser.parse_emotions(emotions, intervals)
        # THEN
        self.assert_equals(
            actual,
            self.utils.read_json_from_resources("intervals-with-emotions.json", True),
        )

    def test_parse_visual_content_moderation(self):
        # GIVEN
        vi_output = self.utils.read_json_from_resources("vi-output.json")
        vcm = vi_output["videos"][0]["insights"]["visualContentModeration"]
        intervals = self.parser.create_intervals(vi_output["videos"][0], "FOO")
        # WHEN
        actual = self.parser.parse_visual_content_moderation(vcm, intervals)
        # THEN
        self.assert_equals(
            actual,
            self.utils.read_json_from_resources(
                "intervals-with-visual-content-moderation.json", True
            ),
        )

    def test_parse_frame_patterns(self):
        # GIVEN
        vi_output = self.utils.read_json_from_resources("vi-output.json")
        frame_patterns = vi_output["videos"][0]["insights"]["framePatterns"]
        intervals = self.parser.create_intervals(vi_output["videos"][0], "FOO")
        # WHEN
        actual = self.parser.parse_frame_patterns(frame_patterns, intervals)
        # THEN
        self.assert_equals(
            actual,
            self.utils.read_json_from_resources(
                "intervals-with-frame-patterns.json", True
            ),
        )


tests/testtimeparser.py
import os

import pytest

from src.parser.timeparser import TimeParser
from tests.testbase import TestBase
from tests.utils import Utils


class TestTimeParser(TestBase):
    """
    This class contains unit tests for UsernameHelper class
    """

    os.environ["MILLISECONDS_INTERVAL"] = "10000"
    time_parser = TimeParser()
    utils = Utils()

    def test_get_related_intervals(self):
        """
        test_alphanum_split for this pattern: "alpha_number_alpha".
        """
        # WHEN
        actual = self.time_parser.get_related_intervals(0, 20000)
        # THEN
        self.assert_equals(actual, [0, 10000])

    def test_get_related_intervals_15_seconds_interval(self):
        """
        test_alphanum_split for this pattern: "alpha_number_alpha".
        """
        # WHEN
        actual = self.time_parser.get_related_intervals(0, 20000)
        # THEN
        self.assert_equals(actual, [0, 10000])

    def test_string_time_to_milliseconds_seconds_only(self):
        """
        test_alphanum_split for this pattern: "alpha_number_alpha".
        """
        # WHEN
        actual = self.time_parser.string_time_to_milliseconds("00:23:00")
        # THEN
        self.assert_equals(actual, 1380000)

    def test_string_time_to_milliseconds_seconds_and_milliseconds(self):
        """
        test_alphanum_split for this pattern: "alpha_number_alpha".
        """
        # WHEN
        actual = self.time_parser.string_time_to_milliseconds("00:23:00.122")
        # THEN
        self.assert_equals(actual, 1380122)

    def test_string_time_to_milliseconds_wrong_time_format(self):
        # WHEN/THEN
        with pytest.raises(Exception) as info:
            self.time_parser.string_time_to_milliseconds("FOO")


src/parser/timeparser.py
import time
from os.path import isfile
from os import getenv

from datetime import datetime
import re

from src.utils.util import Util


class TimeParser:
    def __init__(self):
        """
        This is a constructor for the Time parser.
        It initiates a TimeParser with desired intervals
        e.g. interval_in__milliseconds=10000 creates
        intervals of 10 seconds in that video
        """

        self.interval_in_milliseconds = int(
            Util().config["parser"]["milliseconds-interval"]
            if isfile("config/config.yml")
            else getenv("MILLISECONDS_INTERVAL")
        )

    def get_related_intervals(self, start, end):
        """
        This method returns a list of intervals based on
        start  and end time passed
        e.g.
        start:00:00:00
        end: 00:00:35
        returns [0000,01000,02000,03000] which represents  moments of:
        0 to 10 seconds
        10 to 20 seconds
        20 to 30 seconds
        30 to 40 seconds
        :param start: start time in milliseconds
        :param end: end time in milliseconds
        :return: list of intervals based on start  and end time passed
        """
        intervals = []
        if (
            end - start < self.interval_in_milliseconds
        ):  # CASE: when appearance is within time interval
            intervals.append(
                int(start) - int(start) % self.interval_in_milliseconds
            )  # noqa  E501
        for i in range(int(start), int(end)):
            if i % self.interval_in_milliseconds == 0:
                intervals.append(i)
        return intervals

    @staticmethod
    def string_time_to_milliseconds(string_time):
        """
        This method converts string time to milliseconds
        it uses regex to handle both of the following  cases formats:
        - %H:%M:%S
        - %H:%M:%S.%f
        :param string_time:
        :return: passed time converted to milliseconds
        """
        if re.search(
            "^([0-1]?[0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9]$", string_time
        ):  # noqa: E501,E261
            date_time = datetime.strptime(string_time, "%H:%M:%S")
        elif re.search("^([0-1]?[0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9]\.\d", string_time):
            date_time = datetime.strptime(string_time, "%H:%M:%S.%f")
        else:
            raise ("time format  error")
        date_time = date_time - datetime(1900, 1, 1)
        milliseconds = date_time.total_seconds() * 1000
        return milliseconds

    @staticmethod
    def seconds_to_time_string(seconds):
        """
        This method converts seconds to %H:%M:%S format
        :param seconds:
        :return:
        """

        return str(time.strftime("%H:%M:%S", time.gmtime(seconds)))
